### PR TITLE
[WIP] Property-based testing for snabb config

### PR DESCRIPTION
**Pull request by takikawa**
Initial property-based testing for snabb config, based on code by @kbara.

There are some unimplemented additions that would be useful such as generating `snabb config add` and `snabb config remove` commands. Also, the random value generation doesn't support all YANG types so it should probably be expanded (it supports enough for the snabb/ietf softwire protocols right now).

There are also some minor issues that I think can be fixed later, such as better value generation for list values.

### CODE REVIEW

**Comment by kbara** [comment]
It's looking nice. Some bits, like the crash checks, look fragile but like a reasonably workable approach.

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+local function random_hexes()
+   local str = ""
+   for i=1, 4 do
+      str = str .. random_hex()
+   end
+   return str
+end
+
+-- return a random number, preferring boundary values
+local function choose_range(lo, hi)
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua]: Preferring boundary values is good, but a weighting of 0.5 is probably a bit heavy - how would you feel about 0.1 or 0.2? I'm also inclined to include 0 in the list of numbers to test often.

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+   --   local int64 = value_from_type("int64")
+   --   local exp   = math.random(1, 18)
+   --   return int64 * (10 ^ -exp)
+   elseif prim == "boolean" then
+      return choose({ true, false })
+   elseif prim == "ipv4-address" then
+      return math.random(0, 255) .. "." .. math.random(0, 255) .. "." ..
+             math.random(0, 255) .. "." .. math.random(0, 255)
+   elseif prim == "ipv6-address" then
+      local addr = random_hexes()
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua:220]: This is a fine approach since it's a short string; for longer ones, it makes more sense to build up an array and then use a function that does the whole join in one go.

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+      local len = choose_nat()
+      -- just ascii for now
+      local str = ""
+      for i=0, len do
+         str = str .. string.char(math.random(97, 122))
+      end
+      return str
+   end
+
+   -- TODO: generate these:
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua]: Looking forward to seeing those. :-)

File: src/program/lwaftr/tests/propbased/prop_sameval.lua [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env luajit
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/prop_sameval.lua:1]: Nice property. WDYT about occasionally hammering it with way more than 2 gets/sets, like 100-1000, maybe 1 time in 100?

**Comment by kbara** [comment]
The current version LGTM, though I've flagged a couple of nits.

**Comment by takikawa** [comment]
Note: I've pushed the current state of this branch to the PR. There's one known issue I wanted to fix but didn't quite finish yet, which is that the result of `math.random` in LuaJIT for some of the integer ranges (e.g., uint64) produces non-integers. So sometimes the test generates bad config fragments like `binding-table-version: 1.3115911891594e+16;`.

I plan to fix this by writing and using a random function that has the same interface as the two-argument `math.random` but divides the range in half and randomly selects a half to sample from (since `math.random` is fine on 32-bit ranges).

**Comment by dpino** [comment]
@takikawa So if my understanding is correct the problem is that `math.randon(0, 2^64-1)` sometimes returns a number in float representation (makes sense since Lua numbers are 64-bit doubles), like 1.3115911891594e+16. As `binding-table-version` expects an uint64, would be incorrect to generate a configuration file with the value represented as a floating point value. Would it work to cast that value to an integer using an expression like?

```lua
> = string.format("%0.f", 1.3115911891594e+16)
13115911891594000
```

The result is an string, but if it's for generating a text configuration file, that would work I think.

**Comment by kbara** [comment]
We can't generate the full range of 64 bit ints using a single call to math.random(). As Diego said, we're using LuaJIT with the native backing type being doubles. Your suggested workaround with combining two calls sounds fine to me, as long as you take care with handling the result, making sure it's returning a uint64 rather than a native Lua number and that no intermediate steps using Lua numbers require more precision than a native Lua number can provide.

File: src/program/lwaftr/tests/propbased/common.lua [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+
+function make_handle_prop_args(name, duration, pidbox)
+  local handler = function(prop_args)
+     if #prop_args ~= 1 then
+        print("Usage: snabb quickcheck prop_sameval PCI_ADDR")
+        os.exit(1)
+     end
+  
+     -- TODO: validate the address
+     local pci_addr = prop_args[1]
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/common.lua:15]: Does this code already sufficiently address validation? If so, the TODO comment should be removed, I'd say.

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+
+function run_yang(yang_cmd)
+   local f = io.popen(yang_cmd)
+   local result = f:read("*a")
+   f:close()
+   return result
+end
+
+-- choose a natural number (e.g., index or length of array) by
+-- repeating a cointoss
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua:94]: Why do you want to make it exponentially less likely to choose later options?

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+   elseif prim == "int64" then
+      return choose_range(rng, -9223372036854775809, 9223372036854775807)
+   elseif prim == "uint8" then
+      return choose_range(rng, 0, 255)
+   elseif prim == "uint16" then
+      return choose_range(rng, 0, 65535)
+   elseif prim == "uint32" then
+      return choose_range(rng, 0, 4294967295)
+   elseif prim == "uint64" then
+      return choose_range(rng, 0, 18446744073709551615)
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua]: Wrong numeric type, use a LL suffix.

LuaJIT 2.1.0-beta1 -- Copyright (C) 2005-2015 Mike Pall. http://luajit.org/
JIT: ON SSE2 SSE3 SSE4.1 fold cse dce fwd dse narrow loop abc sink fuse
> print(18446744073709551615 == 18446744073709551614)
true
> print(18446744073709551615LL == 18446744073709551614LL)
false

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+   end
+
+   if prim == "int8" then
+      return choose_range(rng, -128, 127)
+   elseif prim == "int16" then
+      return choose_range(rng, -32768, 32767)
+   elseif prim == "int32" then
+      return choose_range(rng, -2147483648, 2147483647)
+   elseif prim == "int64" then
+      return choose_range(rng, -9223372036854775809, 9223372036854775807)
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua]: Use LL.

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 9]
```diff
@@ -0,0 +1,545 @@
...
+      return choose_range(rng, 0, 255)
+   elseif prim == "uint16" then
+      return choose_range(rng, 0, 65535)
+   elseif prim == "uint32" then
+      return choose_range(rng, 0, 4294967295)
+   elseif prim == "uint64" then
+      return choose_range(rng, 0, 18446744073709551615)
+   -- TODO: account for fraction-digits and range
+   elseif prim == "decimal64" then
+      local int64 = math.random(-9223372036854775809, 9223372036854775807)
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua]: Use LL.

File: src/program/lwaftr/tests/propbased/selftest.sh [hunk 10]
```diff
@@ -0,0 +1,11 @@
+#!/usr/bin/env bash
+
+set -e
+
+SKIPPED_CODE=43
+
+if [ -z $SNABB_PCI0 ]; then exit $SKIPPED_CODE; fi
+
+./snabb quickcheck program.lwaftr.tests.propbased.prop_nocrash $SNABB_PCI0
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/selftest.sh:9]: Ok for now, though ideally I'd like to be able to run these tests without a real NIC at some point, so I can just have my personal build server hammer away at it. Not a blocker for merging.

**Comment by kbara** [comment]
What's the root cause of the test error?

```
src/testlog/program.lwaftr.tests.propbased.selftest.sh:
program/lwaftr/tests/data/icmp_on_fail.conf: loading compiled configuration from program/lwaftr/tests/data/icmp_on_fail.o
program/lwaftr/tests/data/icmp_on_fail.conf: compiled configuration is up to date.
Using time as seed: 19790895
Warning: No NUMA memory affinity.
Pass --cpu to bind to a CPU and its NUMA node.
hugetlb mmap failed (Cannot allocate memory), falling back.
lib/ctable.lua:152: mmap failed: Cannot allocate memory
stack traceback:
	core/main.lua:149: in function <core/main.lua:147>
	[C]: in function 'error'
	lib/ctable.lua:152: in function 'calloc'
	lib/ctable.lua:167: in function 'resize'
	lib/ctable.lua:129: in function 'new'
	apps/lwaftr/ctable_wrapper.lua:41: in function 'initialize_frag_table'
	apps/lwaftr/ipv4_apps.lua:42: in function 'new'
	core/app.lua:305: in function <core/app.lua:304>
	core/app.lua:332: in function 'apply_config_actions'
	apps/config/follower.lua:50: in function 'commit_pending_actions'
	apps/config/follower.lua:62: in function 'handle_actions_from_leader'
	...
	program/lwaftr/setup.lua:507: in function 'run_worker'
	[string "require('program.lwaftr.setup').run_worker({[..."]:1: in function 'f'
	core/main.lua:58: in function <core/main.lua:43>
	[C]: in function 'xpcall'
	core/main.lua:211: in main chunk
	[C]: at 0x00452f60
	[C]: in function 'pcall'
	core/startup.lua:3: in main chunk
	[C]: in function 'require'
	[string "require "core.startup""]:1: in main chunk
100 iterations succeeded.
program/lwaftr/tests/data/icmp_on_fail.conf: loading compiled configuration from program/lwaftr/tests/data/icmp_on_fail.o
program/lwaftr/tests/data/icmp_on_fail.conf: compiled configuration is up to date.
Using time as seed: 23502252
100 iterations succeeded.
program/lwaftr/tests/data/icmp_on_fail.conf: loading compiled configuration from program/lwaftr/tests/data/icmp_on_fail.o
program/lwaftr/tests/data/icmp_on_fail.conf: compiled configuration is up to date.
Using time as seed: 26678293
link report:
Running the same config command twice produced different outputs
Git revision 41a3bcb6f3a64c50d7ceb7aeb1c8554eb4566c67
The property was falsified.
The command was:	./snabb config get -s ietf-softwire 6886 "/softwire-config/description"
Rerun as: snabb quickcheck --seed=26678293 --iterations=11 program.lwaftr.tests.propbased.prop_sameval 0000:82:00.0
EXITCODE: 1
```

**Comment by kbara** [comment]
I currently see two blockers to merging:
a) The numeric issues - those should be straightforward to sort out
b) The test failure - the property based test suite always has the awkward compromise of needing to help us find bugs, but the version that runs in CI should not be expected to fail (ie, it should not trigger known bugs). Making our test suite less deterministic is generally a problem, and intermittent failures are best avoided. (That said, turning up a genuinely new bug at random during CI is fine!)

**Comment by takikawa** [comment]
@kbara re: the error, you mean the quickcheck failure at the bottom rather than the hugetlb error right? It looks like that's after triggering one of those bursts of 100 gets at a time that you suggested (in the previous round of comments), and then causing the lwaftr or something else to get stuck during that. Looking into it.

Update: it may also just be that the timeout is too short on the test.

File: src/program/lwaftr/tests/propbased/genyang.lua [hunk 11]
```diff
@@ -123,6 +124,19 @@ local function random_hexes()
    return str
 end
 
+-- generate a random 64-bit integer
+local function random64()
+   local result = 0
+   local r1 = ffi.cast("uint64_t", math.random(0, 2 ^ 32 - 1))
+   local r2 = ffi.cast("uint64_t", math.random(0, 2 ^ 32 - 1))
+
+   for i=1, 32 do
```
Comment (kbara) [inline src/program/lwaftr/tests/propbased/genyang.lua]: Why do you loop this instead of just multiplying by 4294967296LL?

File: src/program/lwaftr/tests/propbased/genyang.lua (hunk 6 above)
Comment (takikawa) [reply src/program/lwaftr/tests/propbased/genyang.lua:94]: This is used often for generating the `n` in the `[position=n]` part of a `leaf-list` query. Since most of the lists used in the config are small, it's more useful to test smaller numbers to access data that's actually in the config.

Though it's also used for, e.g., picking the length of a random string. So another option would be to just bias, say, 1-5 and then pick a random number from a large range. I could implement that change if it seems better.

File: src/program/lwaftr/tests/propbased/genyang.lua (hunk 11 above)
Comment (takikawa) [reply src/program/lwaftr/tests/propbased/genyang.lua]: Just a brain fart. Pushed a change for that.

File: src/program/lwaftr/tests/propbased/genyang.lua (hunk 6 above)
Comment (kbara) [reply src/program/lwaftr/tests/propbased/genyang.lua:94]: Naw, sounds good as-is, thanks for the clarification.

Comment (kbara) [reply src/program/lwaftr/tests/propbased/genyang.lua:94]: Or specifically, it sounds good as-is with the goal of merging it tomorrow; in the future, I'd like to see a way to generate longer strings with non-trivial probability.

**Comment by kbara** [comment]
LGTM
